src/java.base/share/classes/sun/util/locale/provider/LocaleResources.java

```diff
@@ -531,6 +566,200 @@ public ResourceBundle getJavaTimeFormatData() {
         return rb;
     }
 
+    /**
+     * Returns the actual format pattern string based on the requested template
+     * and calendar type for this locale.
+     *
+     * @param requestedTemplate requested template
+     * @param calType calendar type
+     * @throws IllegalArgumentException if the requested template is invalid
+     * @return format pattern string for this locale, null if not found
+     */
+    public String getLocalizedPattern(String requestedTemplate, String calType) {
+        String pattern;
+        String cacheKey = SKELETON_PATTERN + calType + "." + requestedTemplate;
+
+        removeEmptyReferences();
+        ResourceReference data = cache.get(cacheKey);
+
+        if (data == null || ((pattern = (String) data.get()) == null)) {
+            pattern = getLocalizedPatternImpl(requestedTemplate, calType);
+            cache.put(cacheKey,
+                new ResourceReference(cacheKey, pattern != null ? pattern : "", referenceQueue));
+        } else if ("".equals(pattern)) {
+            // non-existent pattern
+            pattern = null;
+        }
+
+        return pattern;
+    }
+
+    private String getLocalizedPatternImpl(String requestedTemplate, String calType) {
+        initSkeletonIfNeeded();
+
+        // input skeleton substitution
+        var skeleton = substituteInputSkeletons(requestedTemplate);
+
+        // validity check
+        var matcher = VALID_SKELETON_PATTERN.matcher(skeleton);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException("Requested template is invalid: " + requestedTemplate);
```

Comment:
The substitutions are not going to be visible in the exception message.
That might make it harder to identify and fix the cause of the exception. Perhaps the message can make it clear that the matching was done after substitutions and show the 'skeleton'.